lib/tree.rb

```diff
@@ -2,54 +2,130 @@ class TreeNode
+
+    if new_node.key < current_node.key
+      current_node.left = add_helper(current_node.left, new_node)
+    elsif new_node.key > current_node.key
+      current_node.right = add_helper(current_node.right, new_node)
+    end 
+
+    return current_node
+  end 
+
+  # Time Complexity: o(log(n))
+  # Space Complexity: o(1)
+  def add(key, value = nil)
+    new_node = TreeNode.new(key, value)
+    if @root.nil?
+      @root = new_node
+    else
+      add_helper(@root, new_node)
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def find_helper(current_node,key)
+    return nil if current_node.nil? 
+
+    if current_node.key == key
+      return current_node.value
+    elsif key <= current_node.key
+      find_helper(current_node.left, key)
+    else key > current_node.key
+      return find_helper(current_node.right, key)
+    end 
+  end 
+
+
+  # Time Complexity: o(n)
+  # Space Complexity: o(n)
   def find(key)
-    raise NotImplementedError
+    # raise NotImplementedError
+    if @root.nil?
+      return nil
+    else
+      return find_helper(@root, key)
+    end 
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def inorder_helper(current_node, values)
+    return values if current_node.nil?
+
+    inorder_helper(current_node.left, values)
+    values.push({key: current_node.key, value: current_node.value})
+    inorder_helper(current_node.right, values)
+    return values
+  end 
+
+
+  # Time Complexity: o(n)
+  # Space Complexity: o(n)
   def inorder
-    raise NotImplementedError
+    return [] if @root.nil? 
+    return inorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def preorder_helper(current_node, values)
+    return values if current_node.nil?
+
+    values.push({key: current_node.key, value: current_node.value})
+    preorder_helper(current_node.left, values)
+    preorder_helper(current_node.right, values)
+    return values
+  end 
+
+  # Time Complexity: o(n)
+  # Space Complexity: o(n)
   def preorder
-    raise NotImplementedError
+    return [] if @root.nil? 
+    return preorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def postorder_helper(current, values)
+    return values if current.nil?
+
+    postorder_helper(current.left, values)
+    postorder_helper(current.right, values)
+    values.push({key: current.key, value: current.value})
+    return values
+  end
+
+  # Time Complexity:  o(n)
+  # Space Complexity: o(n)
   def postorder
-    raise NotImplementedError
+    return [] if @root.nil?
+    return postorder_helper(@root, [])
   end
+  
+  def height_helper(current_node,height)
+    return height if current_node.nil?
 
-  # Time Complexity: 
-  # Space Complexity: 
+    leftmax = height_helper(current_node.left, height +1 ) 
+    rightmax = height_helper(current_node.right, height + 1)
+
+    return [leftmax, rightmax].max
+  end 
+  
+  # Time Complexity: o(n)
+  # Space Complexity: o(n)
   def height
```

Comment:
👍 The space complexity is O(log n) if the tree is balanced and O(n) if it's unbalanced due to the recursion.  You have the time complexity right.